task_2 на релиз

AlexV0rob opened the pull request:
Добавлен интерфейс главного меню.
Добавлен функционал шахмат, можно играть на одном устройстве.
Для пользователей заведены специальные поля, хранящие режим, в котором бот работает, состояние доски и очерёдность хода.
Шахматы отрисованы с помощью символов, при смене стороны доска поворачивается, есть проверка на возможность хода.

Commit:
Добавление Классов для шахматных фигур

Commit:
Изменена логика MainLogic, добавлен класс User

Добавлены режимы, в которых может работать бот:
0 - главное меню;
1 - режим эхо;
2 - одиночная игра;
MainLogic теперь обрабатывает входящее сообщение, основываясь на режиме,
в котором работает бот;
Класс User реализует сущность пользователя, содержит ID телеграмма,
текущий режим и состояние доски;
Информация о пользователях хранится в ассоциативном массиве

Commit:
Удаление неудачной идеи.
Удаление неудачной идеи с отдельным package для шамхатных
фигру

Commit:
Исправления неправильного использования пакетов

Commit:
Обновление классов шахматных фигур
Обновление классов шахматных фигур без javadoc

Commit:
Добавление функционала игры и ходов

Создана новая сущность MoveButton, помогающая в создании Inline
клавиатуры;
TelegramBot теперь отслеживает callback запросы, формирование сообщений
с приреплением к ним Inline или Reply клавиатур происходит в отдельных
методах;
Добавлен класс MenuHandler, обрабатывающий текстовые сообщения с
управляющими запросами в режиме меню;
Создан класс MoveMaker, отвечающий за обработку callback запросов с
Inline клавиатуры выбора хода и изменение сообщения с текущим ходом.

Commit:
Обновление классов
Изменены все классы, добавлен прототип нового класса, где будет
происходить игра

AlexV0rob committed:
Merge pull request #13 from AlexV0rob/task_1

Из task_1 в task_2

Commit:
Изменение класса пешки
Добавление возможности делать двойной ход из стартовой точки

Commit:
Исправление проверки хода
Теперь также проверяется отсутствие преград на пути
Добавлен новый класс для произведения хода

Commit:
Небольшие изменения
Исправлено название класса пользователя
Изменено условие для switch-case

Commit:
Небольшое изменение првоерки на корректность

Commit:
Добавление Юнит-тестов

Commit:
Небольшая правка в последовательности команд

Commit:
Небольшое упрощение логики коня

Commit:
Обновление Юнит тестов
Небольшое изменение логики

KashKetchup committed:
Merge branch 'task_2' into task_2_Nikita_fix

KashKetchup committed:
Merge pull request #15 from AlexV0rob/task_2_Nikita_fix

Добавление классов фигур

Commit:
Реализация работы с шахматным интерфейсом

В классе MoveMaker добавлена возможность обрабатывать текстовый запрос
хода и отрисовывать доску;
Добавлены тесты класса MoveMaker

Commit:
Merge remote-tracking branch 'origin/task_2' into task_2_Vorobyov

AlexV0rob committed:
Merge pull request #16 from AlexV0rob/task_2_Vorobyov

Добавлен интерфейс шахмат и обработка пользовательского ввода

Commit:
Добавление проверки на шах и мат

Commit:
Добавление тестов для Шаха и мата

KashKetchup committed:
Merge pull request #17 from AlexV0rob/task_2_Nikita_checkmate

Добавление проверки на шах и мат в task_2

Commit:
Объединение логики взаимодействия с пользователем и обработки ходов

Каждый ход, полученный от пользователя в любом виде (текстовое
сообщение, Inline клавиатура) анализируется на возможность. При успешном
результате ход применяется и переходит к другому игроку, иначе ход
необходимо переделать.

AlexV0rob committed:
Merge branch 'task_2' into task_2_fixes_and_union

AlexV0rob committed:
Merge pull request #18 from AlexV0rob/task_2_fixes_and_union

Объединение логики взаимодействия с пользователем и обработки ходов

Commit:
Включение проверки на шах и мат в игровой процесс

Проверка после хода на шаховую и матовую ситуации теперь работает в игре

AlexV0rob committed:
Merge pull request #19 from AlexV0rob/task_2_final_update

Включение проверки на шах и мат в игровой процесс

Commit:
Добавление Javadoc в классы, удаление static методов

Commit:
Добавление Javadoc

AlexV0rob committed:
Merge pull request #20 from AlexV0rob/task_2_checkmate_fix

Добавление Javadoc в классы, удаление static методов

AlexV0rob committed:
Update README.md

Добавлена вторая задача

AlexV0rob committed:
Merge pull request #22 from AlexV0rob/task_2_readme_update

Task 2 readme update

Commit:
Исправление замечаний:
1)Assertions.assertEquals заменены на assertFalse/assertTrue
2)Видоизменён код в некоторых местах для удобства чтения
3)Исправлена критическая ошибка в проверке на существование короля

Commit:
Изменение описания классов

Commit:
Изменение описания методов

Commit:
Исправление ошибки не правильной интерпритацией хода

Commit:
Обновление теста

Commit:
Исправление теста

Commit:
Рефакторинг проверки хода коня

KashKetchup committed:
Merge pull request #23 from AlexV0rob/task_2_checkmate_fix

Исправление недочётов, озвученных на занятии

Commit:
Исправление недочётов, озвученных на занятии:
1) Удалены ненужные @param @return
2) Для итерируемых объектов (типа ArrayList) заменены тесты на
assertIterableEquals
3)Во всех реализациях Chessmen поставлена анотация @Override

KashKetchup committed:
Merge pull request #24 from AlexV0rob/task_2_tests_fix

Исправление недочётов, озвученных на занятии:

vpyzhyanov commented on src/main/java/org/example/Bishop.java:
```diff
@@ -0,0 +1,55 @@
+		  2) Проверяем что слон может так сходить
+		 * */
+		
+		 if (((chessDesk[rawEndPos] == 0 ) || ((chessDesk[rawEndPos] % 2 == 0 ) != isWhite)) && IsThereObstacle(rawStartPos, rawEndPos,chessDesk)){  
```
очень длинная строчка. Перенесите `&& IsThereObstacle(rawStartPos, rawEndPos,chessDesk)` на следующую строку

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Bishop.java:
```diff
@@ -0,0 +1,55 @@
+	 * делением с остатком на 7 проверяется приадлежность хода к левой диагонали
+	 * а делением с остатком на 9 проверяется приадлежность хода к правой диагонали
+	 */
+
```
этот пробел должен быть перед описанием, а не после. Почему у вас JavaDoc улетел от метода?

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Bishop.java:
```diff
@@ -0,0 +1,55 @@
+		 }
+		return false;
+	}
+	/*Проверяем что на пути у нашего слона нету преград
```
оказывается это не JavaDoc!!!!

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Bishop.java:
```diff
@@ -0,0 +1,55 @@
+	 * а делением с остатком на 9 проверяется приадлежность хода к правой диагонали
+	 */
+
+	 public boolean IsThereObstacle(int rawStartPos, int rawEndPos,byte[] chessDesk){
```
методы именуются с маленькой буквы

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Bishop.java:
```diff
@@ -0,0 +1,55 @@
+
+	 public boolean IsThereObstacle(int rawStartPos, int rawEndPos,byte[] chessDesk){
+		if(rawEndPos > rawStartPos) {
+			if((rawEndPos-rawStartPos) % 7 == 0) {
```
все магические числа должны быть вынесены в константы с понятными именами. Никакой магии!

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Bishop.java:
```diff
@@ -0,0 +1,55 @@
+							 return false;
+			}
+			else {
+				 for(int i = rawStartPos + 9; i < rawEndPos; i += 9)
```
здесь и далее - аналогично

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Bishop.java:
```diff
@@ -0,0 +1,55 @@
+		}
+		else {
+			if((rawStartPos-rawEndPos) % 7 == 0) {
+				 if(rawEndPos > rawStartPos){
```
вложенные циклы усложняют код. Тем более в таком простом случае можно просто использовать `&&`

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Castle.java:
```diff
@@ -0,0 +1,49 @@
+		}
+		return false;
+	}
+	public boolean IsThereObstacle(int rawStartPos, int rawEndPos,byte[] chessDesk){
```
1. Нарушена инкапсуляция. `private`
2. Нет JavaDoc

Остальные классы тоже проверьте

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Castle.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Класс для реализации логики перемещения Ладьи
+ */
+class Castle implements  Chessmen{
```
1. Используйте автоформатирование кода при коммите
2. Либо пользуйте fdnjформатированием вручную. в Eclipse сами ищите нужную горячую клавишу

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/CommandHandler.java:
```diff
@@ -0,0 +1,117 @@
+	/**
+	 * Сообщение команды /start
+	 */
+	private final String START_MESSAGE = """
```
все константы должны быть static. Не зря же для них сделано исключение

vpyzhyanov commented on src/main/java/org/example/CommandHandler.java:
```diff
@@ -0,0 +1,117 @@
+	 * Определить тип команды, поменять на соответсвующий режим
+	 * и отправить ответ
+	 */
+	public ArrayList<String> processCommand(String command, User curUser) {
```
Используйте интерфейсы, а не классы, когда это возможно

vpyzhyanov commented on src/main/java/org/example/CommandHandler.java:
```diff
@@ -0,0 +1,117 @@
+		 * 1 - режим эхо
+		 * 2 - одиночная игра (игра на одном устройстве)
+		 */
+		ArrayList<String> messages = new ArrayList<String>();
```
Используйте интерфейсы, а не классы, когда это возможно

vpyzhyanov replied:
Не понял. Зачем вам хранить какие-то сообщения? Выглятит это крайне сомнительным

vpyzhyanov commented on src/main/java/org/example/CommandHandler.java:
```diff
@@ -0,0 +1,117 @@
+		 * При введении некоторых команд переключается режим работы бота
+		 * Список режимов:
+		 * 0 - главное меню
+		 * 1 - режим эхо
```
это надо удалить. В идею вашего бота это не вписывается

vpyzhyanov commented on src/main/java/org/example/CommandHandler.java:
```diff
@@ -0,0 +1,117 @@
+		/*
+		 * При введении некоторых команд переключается режим работы бота
+		 * Список режимов:
+		 * 0 - главное меню
```
Список режимов лучше задать через enum, а не цифрами

vpyzhyanov commented on src/main/java/org/example/CommandHandler.java:
```diff
@@ -0,0 +1,117 @@
+		case "/start":
+			messages.add(START_MESSAGE);
+		case "/menu":
+			curUser.changeMode((byte) 0);
```
Опять магические числа. Притом описание их значения есть в каком-то непонятном месте, которое найти не представляется возможным. Даже в `changeMode` нет этого и ссылки где посмотреть тоже нет. Видимо специально, чтобы никто ничего не понял)))

Используйте enum, тогда всем будет всё понятно.

vpyzhyanov commented on src/main/java/org/example/CommandHandler.java:
```diff
@@ -0,0 +1,117 @@
+			break;
+		case "/help":
+			messages.add(HELP_MESSAGE);
+			break;
```
используйте новый и модный вариант switch с лямбдами

vpyzhyanov replied:
проигнорировали

AlexV0rob replied:
Учитывая переработанный принцип работы класса, не вижу необходимости в использовании switch-выражения

vpyzhyanov replied:
тем не менее вы его использовали

vpyzhyanov commented on src/main/java/org/example/CommandHandler.java:
```diff
@@ -0,0 +1,117 @@
+	 * Определить тип команды, поменять на соответсвующий режим
+	 * и отправить ответ
+	 */
+	public ArrayList<String> processCommand(String command, User curUser) {
```
Почему вдруг возвращается список? Это что для того, чтобы построчно выводить? Не надо так, используйте многострочную строку, для перевода на следующую строку используйте специальный служебный символ `\n`

vpyzhyanov replied:
Нет ответа!!!

AlexV0rob replied:
Причиной является то, что бот общается с человеком: пользовательский интерфейс выглядит приятнее при выводе приветственных и переходных сообщений, которые нужно вывести только один раз при переходе из одного состояния (например, меню) к другому (например, игра). В случае с Телеграмом есть и практическая причина: Телеграм не позволяет добавлять к одному сообщению более одного набора кнопок (будь то их отсутствие, Reply или Inline клавиатуры), а при переходе с одного вида на другой автоматически их не заменяет, поэтому, чтобы убрать одну клавиатуру и вывести другую, есть только два способа: отправлять сообщение, очищающее клавиатуру, а потом его сразу же его удалять, или выводить промежуточное сообщение, очищающее клавиатуру, а потом уже основное с необходимыми кнопками.

vpyzhyanov replied:
> Телеграм не позволяет добавлять к одному сообщению более одного набора кнопок (будь то их отсутствие, Reply или Inline клавиатуры)

Понятное дело, это и не нужно, в этом нет смысла.

> а при переходе с одного вида на другой автоматически их не заменяет, поэтому, чтобы убрать одну клавиатуру и вывести другую, есть только два способа: отправлять сообщение, очищающее клавиатуру, а потом его сразу же его удалять, или выводить промежуточное сообщение, очищающее клавиатуру, а потом уже основное с необходимыми кнопками

Никогда такие костыли не были нужны, а сейчас вдруг необходимы. Это очень жёсткий костыль. Вероятно он вызван неправильной работа с интерфейсом телеграмма. **В следующей задаче этот костыль необходимо убрать!** И естественно метод processCommand должен возвращать только один ответ

KashKetchup replied:
Будет исправлено

vpyzhyanov commented on src/main/java/org/example/EchoHandler.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Реализует функцию эхо
+ */
+public class EchoHandler {
```
удалить

vpyzhyanov commented on src/main/java/org/example/GameHandler.java:
```diff
@@ -0,0 +1,98 @@
+		byte[] curDesk = user.getBoard();
+		switch((figureCode + 1) / 2)
+		{
+		case 1:
```
везде магические числа

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameHandler.java:
```diff
@@ -0,0 +1,98 @@
+	public byte ProgressHandler(User user, int rawStartPos, int rawEndPos, byte figureCode){
+		Chessmen lastChessmen = null;
+		boolean isNormalMove = false;
+		if((rawStartPos > 63 || rawStartPos < 0 ) || (rawEndPos > 63 || rawEndPos < 0))
```
везде магические числа

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameHandler.java:
```diff
@@ -0,0 +1,98 @@
+ * Класс для произведения хода у конкретного пользователя
+ */
+public class GameHandler{
+	public byte ProgressHandler(User user, int rawStartPos, int rawEndPos, byte figureCode){
```
1. нет JavaDoc
2. Особенно непонятно возвращаемое значение
3. методы пишутся с маленькой буквы

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameHandler.java:
```diff
@@ -0,0 +1,98 @@
+	/**
+	 * Проверка на рубку короля
+	 */
+	public boolean 	IsThisMoveOnKing(byte[]curBoard,int rawPosition, boolean isWhiteMove){
```
используйте автоформатирование. Тут и лишние пробелы есть и отсутствие их

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameHandler.java:
```diff
@@ -0,0 +1,98 @@
+	 */
+	public boolean 	IsThisMoveOnKing(byte[]curBoard,int rawPosition, boolean isWhiteMove){
+		if(isWhiteMove)
+			return curBoard[rawPosition] == 11;
```
везде магические числа

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameHandler.java:
```diff
@@ -0,0 +1,98 @@
+			lastChessmen = pawn;
+			break;
+		case 2:
+			Chessmen castle = new Castle();
```
неужели вы каждый раз создаёте фигуры заново? Память лучше бы пожалели

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameHandler.java:
```diff
@@ -0,0 +1,98 @@
+	{
+		int i = 0 ;
+		if(isWhiteMove){
+			for(; i < 64; ++i)
```
используйте while

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameHandler.java:
```diff
@@ -0,0 +1,98 @@
+				return true;
+		}
+		else{
+			for(; i < 64; ++i)
```
используйте while

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Main.java:
```diff
@@ -5,6 +5,7 @@
 /**
  * Основной класс, где происходит запуск бота
  */
+
```
зачем добавлять лишнюю строчку?

vpyzhyanov commented on src/main/java/org/example/MainLogic.java:
```diff
@@ -1,60 +1,43 @@
+			/*
+			 * Коды режимов:
+			 * 0 - главное меню
+			 * 1 - режим эхо
```
второй раз одно и то же пишете. Когда вы делаете одно и то же второй раз, это повод задуматься

vpyzhyanov commented on src/main/java/org/example/MainLogic.java:
```diff
@@ -1,60 +1,43 @@
-			return echoMessage(userInput);
+			//Если режим другой, то произошла ошибка, выходим в главное меню
+			default:
+				responses = new CommandHandler().processCommand("/menu", curUser);
```
В целом этот класс получился довольно хорошо с точки зрения ООП. Правда было бы здорово, если бы в нём не было дублирования, магических числел и постоянного создания обработчиков команды (их можно создать один раз и использовать несколько раз).

vpyzhyanov commented on src/main/java/org/example/MenuHandler.java:
```diff
@@ -0,0 +1,49 @@
+	/**
+	 * Сообщение для включения режима эхо
+	 */
+	private final String ECHO = "Включить эхо-мод";
```
static

vpyzhyanov commented on src/main/java/org/example/MenuHandler.java:
```diff
@@ -0,0 +1,49 @@
+		default:
+			ArrayList<String> responses = new CommandHandler()
+				.processCommand("/menu", curUser);
+			responses.add(0, DIDNT_UNDERSTAND);
```
Вот вы и попались. Сейчас вы ощутите то, почему надо использовать интерфейсы. ArrayList неэффективен в данном случае, поэтому замените везде ArrayList на LinkedList, т.к. вставка в начало в нём значительно быстрее.

vpyzhyanov commented on src/main/java/org/example/MoveMaker.java:
```diff
@@ -0,0 +1,314 @@
+	/**
+	 * Определить, на какой стадии находится ход, и отдать на обработку
+	 */
+	public ArrayList<MoveButton> assembleMove(String callbackText, User curUser) {
```
опредилить на какой стадии находится код?? То есть вы хотите сказать, что вы не храните состояние игры? Зато храните какие-то непонятные сообщения, создаёте фигуры каждый раз заново. В чём проблема хранить состояние?

В общем так дело не пойдёт. Вся игра у вас должна хранить состояние, вы должны всегда знать где какие фигруры находятся. Чтобы никаких "определить" у вас не было.

vpyzhyanov commented on src/main/java/org/example/MoveMaker.java:
```diff
@@ -0,0 +1,314 @@
+		}
+		
+		//Добавление нового текста сообщения, ставится в позицию 0
+		newButtons.add(0, new MoveButton(messageType, newMessage));
```
тоже неэффективная коллекция.

В целом несколько странной и неочевидной видится идея добавлять что-то в начало.

vpyzhyanov commented on src/main/java/org/example/MoveMaker.java:
```diff
@@ -0,0 +1,314 @@
+	 * Проверить, какие фигуры остались на доске
+	 */
+	public ArrayList<MoveButton> whichFiguresLeft(User curUser) {
+		ArrayList<MoveButton> leftFigures = new ArrayList<MoveButton>();
```
и тут тоже надо использовать интерфейсы. Везде

vpyzhyanov commented on src/main/java/org/example/MoveMaker.java:
```diff
@@ -0,0 +1,314 @@
+			}
+		}
+		
+		for (int i = 0; i < 8; ++i) {
```
наши любимые магические константы

vpyzhyanov commented on src/main/java/org/example/MoveMaker.java:
```diff
@@ -0,0 +1,314 @@
+	/**
+	 * Получить запрос на передвижение фигуры из пользовательского ввода
+	 */
+	public ArrayList<String> fromStringToQuery(String messageText, User curUser) {
```
назовите метод в соответствии того, что он делает. `parseMoveCommand`. Вот только возвращаемое значение никак не вписывается в это описание метода. "Запрос" не может быть в виде ArrayList<String>

Возможно тут должен возвращаться класс, который указывает какую фигуру и куда передвинуть. Правда непонятно зачем фигура в таком "запросе", ведь по позиции и так понятно какая это фигура.

vpyzhyanov commented on src/main/java/org/example/MoveMaker.java:
```diff
@@ -0,0 +1,314 @@
+			}
+		}
+		
+		String query = "move_" + ((char) figureCode) + "_" + ((char) startPosition) +
```
вот оно что. То есть "запрос" - это строка. Убираем это. Чтобы "запрос" был классом, а не строкой непонятного формата. У нас всё-таки ООП. 

Я бы ещё рекомендовал заменить слово "запрос" на слово "ход", чтобы хоть как-то его приблизить к действительности, к шахматам

vpyzhyanov commented on src/main/java/org/example/MoveMaker.java:
```diff
@@ -0,0 +1,314 @@
+		String[] messages = new String[3];
+        int buttonsCount = buttons.size();
+        boolean hasError = false;
+        String errorMessage = "";
```
почему вдруг класс, делающий ход пишет какие-то сообщения пользователю? Он не должен так делать, его задача сделать ход. Если ход невозможен - бросайте исключение

vpyzhyanov commented on src/main/java/org/example/MoveMaker.java:
```diff
@@ -0,0 +1,314 @@
+	/**
+	 * Нарисовать текущее состояние доски
+	 */
+	public String printBoard(User curUser) {
```
этот метод никак не может относится к классу "MoveMaker", т.к. к передвижению он никак не относится. Это к вопросу про хранение состояния игры, на самом деле. В идеале должен быть класс, который хранит состояние и умеет отрисовать текущее состояние игры в виде строки. А класс по передвижению по идее вообще не должен существовать, т.к. передвижение - это изменение состояния игры. Максимум могут существовать классы, которые описываю модель передвижения отдельных фигур. Они, кстати, у вас есть и это очень хорошо

vpyzhyanov commented on src/main/java/org/example/Position.java:
```diff
@@ -0,0 +1,12 @@
+package org.example;
+
+public class Position
```
нет JavaDoc

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Position.java:
```diff
@@ -0,0 +1,12 @@
+
+public class Position
+{
+	int line;
```
класс ущемляет права инкапсуляции в этом мире. Его засудят)))

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Position.java:
```diff
@@ -0,0 +1,12 @@
+	public Position(int pos)
+	{
+		this.line = pos / 8;
+		this.column = pos % 8;
```
смысл этого конструктора мне непонятен, как и смысл `pos`. Судя по этим операциям творится какая-то магия

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -29,27 +41,275 @@ public class TelegramBot implements LongPollingSingleThreadUpdateConsumer {
+            }
+            User curUser = users.get(chatId);
+            
+            //Получем тексты ответов и создаём из них ответные сообщения
```
1. О_о то есть получаете ответы и начинаете их редактировать? А смысл? Может сразу сформировать нормальные ответы?
2. Ответственность этого класс просто отправить сообщение в телеграмм, он не должен формировать какие-то сообщения

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -29,27 +41,275 @@ public class TelegramBot implements LongPollingSingleThreadUpdateConsumer {
+            
+            String[] texts = new String[3];
+            int buttonsCount = newButtons.size();
+            boolean hasError = false;
```
1. Такие флажки - признак кривой архитектуры. Флажки - всегда признак кривой архитектуры и каких-то костылей
2. А ещё они намекают на то, что вы тут начали определять логику, хотя логика должна быть точно не здесь

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -29,27 +41,275 @@ public class TelegramBot implements LongPollingSingleThreadUpdateConsumer {
+            	if (newButtons.get(i).getCallbackQuery().equals("side")) {
+            		texts[0] = newButtons.get(i).getButtonMessage();
+            		newButtons.remove(i);
+            	} else if (newButtons.get(i).getCallbackQuery().equals("board")) {
```
действительно, сюда проникла логика

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -29,27 +41,275 @@ public class TelegramBot implements LongPollingSingleThreadUpdateConsumer {
+    	 * 2 - одиночная игра (игра на одном устройстве)
+    	 */
+    	case 0:
+    		return buildMenuMessages(chatId, texts);
```
ого, очень много логики сюда перекочевало. Вся цель предыдущей задачи на смарку. Мы старались разделять логику, но всё напрасно, опять скатились туда же

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -29,27 +41,275 @@ public class TelegramBot implements LongPollingSingleThreadUpdateConsumer {
+               	.replyMarkup(ReplyKeyboardMarkup
+                           .builder()
+                           .keyboardRow(new KeyboardRow(
+                          		 "Включить эхо-мод",
```
кнопки тоже должны определяться в логике. Тут лишь они должны только создаваться, т.к. созданием управляет телеграмм

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -29,27 +41,275 @@ public class TelegramBot implements LongPollingSingleThreadUpdateConsumer {
+    	 * остатки совпадают, приоритет количества кнопок 2 -> 3 -> 5 -> 4.
+    	 */
+    	int buttonsCount = buttons.size();
+    	int remOf2 = buttonsCount % 2;
```
опять какая-то магия. Впрочем эта магия похоже полезная

vpyzhyanov commented on src/main/java/org/example/User.java:
```diff
@@ -0,0 +1,92 @@
+ * Сущность пользователя, хранит информацию о текущем состоянии чата
+ * с конкретным пользователем
+ */
+public class User {
```
судя по полям, этот класс хранит вовсе не чат. И не ялвяется поьзователем

vpyzhyanov commented on src/main/java/org/example/User.java:
```diff
@@ -0,0 +1,92 @@
+	/**
+	 * Узнать, кто сейчас ходит
+	 */
+	public boolean doesWhitesMove() {
```
этот метод прямо доказывает, что не про пользователя речь

vpyzhyanov commented on src/test/java/org/example/ChessmenTest.java:
```diff
@@ -0,0 +1,119 @@
+  			  4, 6, 8, 10, 12, 8, 6, 4};
+    	Chessmen knight = new Knight();
+    	Assertions.assertFalse(knight.checkMove(1,17,board,false));
+    	Assertions.assertTrue(knight.checkMove(1,11,board,false));
```
по 2 проверки на каждую фигуру как-то маловато, чтобы судить о корректности хода. Их должно быть как минимум 4.

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/test/java/org/example/MenuHandlerTest.java:
```diff
@@ -0,0 +1,121 @@
+	 * Сообщение для начала новой однопользовательской игры
+	 * (игры на одном устройстве)
+	 */
+	private final String SINGLEGAME = "Начать новую игру";
```
1. не вижу ни одной проверки с использованием этих значений
2. Вижу дублирование логики основного приложения. Это обесценивает весь тест целиком.
3. Не увидел ни одной проверки самого меню, что было заявлено в описании теста

vpyzhyanov reviewed (changes requested):
очень много замечаний. Самое страшное то, что потеряна цель предыдущей задачи, где мы старались разделить логику бота и телеграмма. Теперь это всё насмарку.

Commit:
Работа над ошибками:
1) В классах фигур добавлены JavaDoc'и для методов проверки
2) Переделан класс Position
3) Добавлен подобрый JavaDoc в GameHandler
4) Изменены Junit тесты
5) Изменен GameHanlder, теперь фигуры создаются только 1 раз, при
создании экземпляра класса GameHanlder
6)Добавлен JavaDoc для GameHanlder

Commit:
Изменение архитектуры бота

Полностью переработана архитектура и логика работы бота: TelegramBot
теперь отвечает только за работу с Telegram, убран режим эхо, MainLogic
играет большую роль в управлении пользовательским вводом, добавлен класс
GameState, хранящий состояние игры и формирующий кнопки и сообщение с
состоянием доски

Commit:
Решение Pull Request из task_2_tests_fix в task_2_big_fix

Commit:
Улучшение архитектуры бота, добавление кнопок

Добавлена возможность вывода Inline и Reply клавиатур, кнопки для
которых сначала формируются в KeyboardCreator, а потом преобразуются в
понимаемые Телеграмом структуры в TelegramBot.
Теперь методы обработки ввода возвращают списки сообщений на отправку
для возможности вывода промежуточных сообщений перехода между режимами,
которые скрывают клавиатуры предыдущего для корректного отображения
новых.
Добавлены тесты на некоторые классы.

Commit:
Исправления тестов для шахматных фигур

AlexV0rob committed:
Merge pull request #27 from AlexV0rob/task_2_big_fixes_union

Task 2 big fix моя часть

KashKetchup committed:
Merge branch 'task_2_big_fixes_union' into task_2_gameHandler_fix

KashKetchup committed:
Merge pull request #28 from AlexV0rob/task_2_gameHandler_fix

Моя часть в Big Fix

Commit:
Расширены тесты для шахматных фигур
Создана функция для отбора всех доступных ходох для фигуры

KashKetchup committed:
Merge pull request #29 from AlexV0rob/task_2_local_fixes

Расширены тесты для шахматных фигур

Commit:
Исправления модификатора доступа у констант

KashKetchup committed:
Merge pull request #30 from AlexV0rob/task_2_local_fixes

Исправления модификатора доступа у констант

Commit:
Добавление тестов

Добавлены тесты для классов MoveConstructor, KeyboardCreator, GameState,
MainLogic

AlexV0rob committed:
Merge pull request #31 from AlexV0rob/task_2_big_fix_new_tests

Добавление тестов

KashKetchup committed:
Изменение Логики с помощью Position Conerter

Commit:
Изменение всех классов под positionConverter

KashKetchup committed:
Изменение всех классов фигур под
PositionConverter!
Также переработаны everyPossibleMoves для всех фигур

KashKetchup committed:
Исправление критических ошибок в логике шахматных фигур

KashKetchup committed:
Изменение классов для verticalMove
HorizontalMove

KashKetchup committed:
Исправление невозможности завершить игру, исправление ошибки проверки цвета в Rook.java и GameHandler

KashKetchup committed:
Добавление тестов для gameHandler

Commit:
Исправление MoveConstructorTest

AlexV0rob committed:
Merge pull request #32 from AlexV0rob/task_2_MoveConstructorTest_fix

Исправление MoveConstructorTest

KashKetchup committed:
Изменение тестов для gameHandler

KashKetchup committed:
Исправление ошибки в неправильном подсчёте позиции

KashKetchup committed:
Merge remote-tracking branch 'origin/task_2_big_fix_new_tests' into task_2_big_fix

# Conflicts:
#	src/main/java/org/example/Bishop.java
#	src/main/java/org/example/MoveConstructor.java
#	src/main/java/org/example/Pawn.java
#	src/main/java/org/example/Position.java
#	src/main/java/org/example/Rook.java

KashKetchup committed:
Удаление лишний полей в Bishop.java

Commit:
Замена LinkedList на ArrayList в CommandHandler

AlexV0rob committed:
Merge pull request #35 from AlexV0rob/task_2_LinkedList_fix

Замена LinkedList на ArrayList в CommandHandler

KashKetchup committed:
Замена List на ArrayList

KashKetchup committed:
Merge remote-tracking branch 'origin/task_2_big_fix' into task_2_big_fix

KashKetchup committed:
Merge pull request #34 from AlexV0rob/task_2_big_fix

Финальное исправление task_2

KashKetchup committed:
Исправление неправильного массива доски

vpyzhyanov commented on src/main/java/org/example/MoveConstructor.java:
```diff
@@ -0,0 +1,230 @@
+    /**
+     * Возможный состояния готовности хода
+     */
+    public static enum STATUS {
```
static запрещён!

AlexV0rob replied:
Здравствуйте! Уточняю, это же, по идее, константы, у которых нужно static добавлять. Или к enum это не относится, потому что его действие выходит за рамки одного класса?

KashKetchup replied:
Исправлено

vpyzhyanov replied:
enum - это как класс. Он константой не является, он даже переменной/полем не является. В регламенте есть разрешено использовать только для конкретных типов переменных, для всех остальных - запрещено

vpyzhyanov commented on src/main/java/org/example/PositionConverter.java:
```diff
@@ -0,0 +1,139 @@
+
+    /**
+     *
+     */
```
пусто

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/PositionConverter.java:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    /**
+     * с
```
пусто

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/PositionConverter.java:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * направление движения фигуры
+     */
+    public static enum DIRECTION_OF_SHIFT {
```
static запрещён!

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/PositionConverter.java:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * варианты расположеиния позиций между друг другом
+     */
+    public static enum SHIFT_PROPERTY {
```
static запрещён!

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/PositionConverter.java:
```diff
@@ -0,0 +1,139 @@
+package org.example;
+
+public class PositionConverter {
```
Нет JavaDoc!

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/Chessmen.java:
```diff
@@ -1,10 +1,20 @@
+    /**
+     * Все возможные ходы данной фигуры с данной позиции
+     */
+    public ArrayList<Integer> everyPossibleMove(int rawStartPos, byte[] chessDesk,
```
Используйте интерфейсы, а не классы

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameState.java:
```diff
@@ -0,0 +1,359 @@
+    /**
+     * Получить список кнопок для следующей части хода
+     */
+    public List<IdentificatedButton> nextButtons() {
```
На развитие архитектуры: состояние игры должно управлять только состоянием игры, а не заниматься управлением кнопками. Кнопки должны создаваться в том классе, который отвечает за взаимодействие с пользователем

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameState.java:
```diff
@@ -0,0 +1,359 @@
+     *
+     * @return ход в текстовом представлении, которое поймёт MainLogic
+     */
+    public String assembleMove() {
```
А тут нарушена инкапсуляция. Я так понял это не сам ход, а лишь его результат. И тут нет никаких параметров для хода. Настоящая инкапсуляция делает ход и получает результат в одном методе. То есть в хорошем методе ход не делится на части (по крайней мере публичные части)

AlexV0rob replied:
Нет, это не результат хода. Так как при использовании inline клавиатуры для совершения хода идёт его разделение на три части (выбор фигуры, выбор начальной позиции, выбор конечной позиции), после полной готовности нужно ход собрать и передать на реализацию. Этот метод только превращает значения трёх переменных в одну строку, которую проще передавать внутри программы.

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameState.java:
```diff
@@ -0,0 +1,359 @@
+    /**
+     * Собрать текущий ход
+     *
+     * @return ход в текстовом представлении, которое поймёт MainLogic
```
то, что в этом методе что-то пошло не так я заметил сначала именно по этой строчке. Какая-то завязка на внешний класс. Классы нижнего уровня ничего не должны знать про классы верхнего уровня.

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameState.java:
```diff
@@ -0,0 +1,359 @@
+                
+                %s
+                %s %s %s %s
+                """.formatted(side,
```
Вы наверняка должны были заметить, что это класс довольно много строк кода занимает, слишком много. А ещё он делает куда больше, чем просто отвечает за состояние. В частности он занимается подготовкой доски к выводу в бота. Это не его ответственность, это надо убрать в отдельный класс.

vpyzhyanov commented on src/main/java/org/example/GameState.java:
```diff
@@ -0,0 +1,359 @@
+     * Установить состояние доски из внешнего массива кодов фигур
+     * (не рекомендуется для совершения ходов)
+     */
+    public void setBoard(byte[] newBoard) {
```
Открыли портал в мир без инкапсуляции и страдания. Самое страшное, что это вы используете только для тестов. Надо закрыть к нему доступ через private.

> (не рекомендуется для совершения ходов)

Тогда и эти приписки не надо будет делать.

В тестах можете задавать состояние доски через конструктор.

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameState.java:
```diff
@@ -0,0 +1,359 @@
+    public void setBoard(byte[] newBoard) {
+        int newBoardSquaresCount = newBoard.length;
+        int i = 0;
+        for (; i < SQUARES_COUNT && i < newBoardSquaresCount; ++i) {
```
либо используйте нормальный цикл for, либо используйте while

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameState.java:
```diff
@@ -0,0 +1,359 @@
+        for (; i < SQUARES_COUNT && i < newBoardSquaresCount; ++i) {
+            chessboard[i] = newBoard[i];
+        }
+        for (; i < SQUARES_COUNT; ++i) {
```
и тут

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -5,34 +5,51 @@
+	 * Ассоциативный массив с парами (ID чата телеграм, состояние игры)
 	 */
-	private HashMap<Long, User> users = new HashMap<>();
+	private HashMap<Long, GameState> games = new HashMap<Long, GameState>();
```
Используйте интерфейсы

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -46,270 +63,184 @@ public TelegramBot(String botToken) {
+        	keyboardsForMessages.add(new ReplyKeyboardRemove(true));
+        }
+        
+        Iterator<String> responseMessages = responseMessagesTexts.iterator();
```
Прошу не путать коллекции с итератором. Поэтому название переменной должно быть `responseMessagesTextsIterator`

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -46,270 +63,184 @@ public TelegramBot(String botToken) {
+        }
+        
+        Iterator<String> responseMessages = responseMessagesTexts.iterator();
+        Iterator<ReplyKeyboard> messagesKeyboards = keyboardsForMessages.iterator();
```
И тут

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/GameHandler.java:
```diff
@@ -1,98 +1,142 @@
+    /**
+     * Является ли ход шахом
+     */
+    private boolean isCheck(int rawStartPos,
```
название методы и его описание из разных опер. Название звучит как "проверка?" Это общепринятое слово и ему требуется контекст, лучше назвать `isCheckMove`

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/MainLogic.java:
```diff
@@ -1,43 +1,145 @@
+     * Скомпилированное регулярное выражение, соответствующее полностью
+     * введённому ходу в текстовом виде
+     */
+    private final static Pattern MOVE_PATTERN =
```
static запрещён!

AlexV0rob replied:
В этом классе static использован к чистым константам, объясните, пожалуйста, почему здесь его использовать нельзя.

KashKetchup replied:
Исправлено

vpyzhyanov replied:
> В этом классе static использован к чистым константам, объясните, пожалуйста, почему здесь его использовать нельзя.

Потому что у нас в регламенте это написано. И вы не спрашивали разрешения использования static для этой конкретной ситуации. По крайней мере это нигде не зафиксировано. В переписке не нашёл

Тем не менее, это объект является неизменяемым, поэтому он может быть статичным, то есть он является константой

vpyzhyanov commented on src/main/java/org/example/MainLogic.java:
```diff
@@ -1,43 +1,145 @@
+    /**
+     * Какая-то часть хода в виде callback запроса
+     */
+    private final static Pattern CALLBACK_PATTERN =
```
static запрещён!

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -46,270 +63,184 @@ public TelegramBot(String botToken) {
+    	if (!games.containsKey(chatId)) {
+        	games.put(chatId, new GameState());
+        }
+        GameState currentGameState = games.get(chatId);
```
Ещё одно нарушение ответственности классов. Управление игрой - это часть логики, а не телеграмма. Телеграмм вообще не должен догадываться о том, что у вас там есть какие-то игры

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/test/java/org/example/PositionConverterTest.java:
```diff
@@ -0,0 +1,59 @@
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class PositionConverterTest {
```
нет JavaDoc

KashKetchup replied:
Исправлено

vpyzhyanov reviewed (changes requested):
Уже существенно лучше, но если не выполнить базовые требования, то сдать задачу не получится

Commit:
Переработка архитектуры бота (безумие - это точное повторение одного и..

Работа классов в целом не изменилась, но некоторые функции были вынесены
в отдельные классы.
Группа классов *State теперь отвечает за различные состояния и не
содержит ничего лишнего.
Доска была переведена из одномерного вида в двумерный, что позволило
сделать более логичными перемещения по доске, дле обозначения позиции
добавлена запись PositionOnBoard, PositionConverter удалён за
ненадобностью.
TelegramBot больше не содержит состояния пользователей, теперь это
задача MainLogic.
Интерфейс InputHandler и реализующего его группа классов *InputHandler
реализуют управление вводом в различных режимах.
Удален все старые классы и их тесты за ненадобностью.
Различные исправления и улучшения производительности.

Commit:
Перемещение файлов в директорию java

KashKetchup committed:
Изменение тестов для фигур

Commit:
Добавление пакетов, тестов, JavaDoc, кнопок и режимов

Сообщения пользователю теперь передаются в виде списка с добавление к
каждому соответствующей клавиатуры.
Добавлено управление ходом с помощью кнопок.
Добавлены режимы пользователя, обработка ввода теперь производится по
ним.
Немного изменена логика проверки хода для пешки.
Для всех классов добавлены тесты.
Для всех классов и тестов добавлены JavaDoc.
Общие по смыслу классы перемещены в соответствующие пакеты.
Различные улучшения производительности.

KashKetchup committed:
Добавление обновлённой версии allPossibleMoves для всех фигур

AlexV0rob committed:
Merge pull request #36 from AlexV0rob/task_2_new_architecture_Vorobyov

Добавление пакетов, тестов, JavaDoc, кнопок и режимов

KashKetchup committed:
Merge remote-tracking branch 'origin/task_2_new_Chessmen_test' into task_2_architecture_reworking

# Conflicts:
#	src/main/java/org/example/chess/Bishop.java
#	src/main/java/org/example/chess/Chessmen.java
#	src/main/java/org/example/chess/GameHandler.java
#	src/main/java/org/example/chess/King.java
#	src/main/java/org/example/chess/Knight.java
#	src/main/java/org/example/chess/Pawn.java
#	src/main/java/org/example/chess/Queen.java
#	src/main/java/org/example/chess/Rook.java

KashKetchup committed:
Добавление обновлённой версии allPossibleMoves для всех фигур

KashKetchup committed:
Небольшое зименнеие javaDoc'а для коня

KashKetchup committed:
Добавление JavaDoc для MIN/MAX_SIDE_VALUE И processMove

KashKetchup committed:
Изменение isCheckMove

KashKetchup committed:
Исправление критической ошибки в isCheckMove

KashKetchup committed:
Добавление нормального JavaDoc для GameHandler

KashKetchup committed:
Добавление  JavaDoc для moveProperty

KashKetchup committed:
Удаление ненужного static

Commit:
Небольшие исправления в логике шаха и пешки

Commit:
Изменение логики работы InGameInputHandler, исправление проверки на мат

InGameInputHandler теперь управляет состояние игры пользователя, а не
MoveHandler.
Проверка на мат теперь работает исправно.
MovePartsHandler переименован в MoveHandler в связи с урезанием
функционала.

KashKetchup committed:
Удаление ненужного static

KashKetchup committed:
Merge remote-tracking branch 'origin/task_2_architecture_reworking' into task_2_architecture_reworking

KashKetchup committed:
Merge pull request #37 from AlexV0rob/task_2_architecture_reworking

Обновление Архитектуры Бота

KashKetchup committed:
Удаление некорректного Static

KashKetchup committed:
Возвращение MateMove в рабочее состояние

KashKetchup committed:
Merge pull request #38 from AlexV0rob/task_2_architecture_reworking

Удаление некорректного Static

AlexV0rob commented on src/main/java/org/example/chess/GameHandler.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    /**
+     *
```
Добавь Javadoc

AlexV0rob commented on src/main/java/org/example/inputHandlers/CommandHandler.java:
```diff
@@ -0,0 +1,91 @@
+ * Обработчик хода фигуры
+ */
+public class CommandHandler {
+    private final GameTranslator gameTranslator = new GameTranslator();
```
Добавь Javadoc

AlexV0rob commented on src/test/java/org/example/ButtonsCreatorTest.java:
```diff
@@ -0,0 +1,89 @@
+		Assertions.assertEquals(pawnMoves.size(), currentButtons.size());
+		moveState.nextStatus("d5");
+		currentButtons = buttonsCreator.getGameButtons(moveState, BOARD, false);
+		Assertions.assertIterableEquals(List.of(), 
```
Поставь во второй параметр currentButtons

KashKetchup committed:
Добавление javaDoc'ов

KashKetchup committed:
Merge pull request #39 from AlexV0rob/task_2_architecture_reworking

Добавление недостающих javaDoc'ов

Commit:
Изменены тесты для GameTranslator на более однозначные

AlexV0rob committed:
Merge pull request #40 from AlexV0rob/task_2_small_test_fix

Изменены тесты для GameTranslator на более однозначные

AlexV0rob commented:
Список изменений и исправлений:
У всех enum убраны модификаторы static.
Доска теперь является двумерным массивом, для обозначения позиции введена запись PositionOnBoard вместо класса PositionConverter.
Ответственность и методы класса GameState разнесены в несколько классов: GameTranslator, MoveState, GameState ButtonsCreator. Сам GameState теперь содержит только состояние игры, доска задаётся через конструктор.
Изменена архитектура бота, но незначительно: более чётко определена ответственность классов.

AlexV0rob commented on src/main/java/org/example/chess/Queen.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * слон для проверки диагональных ходов
+     */
+    private static Chessmen BISHOP;
```
Убери static и добавь final

KashKetchup replied:
Исправлено

AlexV0rob commented on src/main/java/org/example/chess/Queen.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * ладья для проверки вертикальных и горизонтальных ходов
+     */
+    private static Chessmen ROOK;
```
Убери static и добавь final

KashKetchup replied:
Исправлено

KashKetchup committed:
Удаление ненужного static

KashKetchup committed:
Добавление final в Queen.java

KashKetchup committed:
Merge pull request #42 from AlexV0rob/task_2_architecture_reworking_Queen_fix

Удаление ненужного static в Queen.java

Commit:
Избавление от static в MovePartsConverter

AlexV0rob committed:
Merge pull request #43 from AlexV0rob/task_2_MovePartsConverter_fix

Избавление от static в MovePartsConverter

vpyzhyanov commented on src/main/java/org/example/chess/Bishop.java:
```diff
@@ -0,0 +1,191 @@
+    /**
+     * Расположение позиций друг относительно друга
+     */
+    private enum positionRelatives {
```
enum как и класс должен называться с заглавной буквы

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/ButtonsCreator.java:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * Список фигур для получения всех возможных ходов
+     */
+    private final static Chessmen[] FIGURES = {
```
static запрещён!

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/states/UserState.java:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * Начальная доска
+     */
+    private final static byte[][] START_BOARD =
```
static запрещён! Вообще-то массив не является неизменяемым объектом поэтому константой быть не может. Т.е. можно спокойно сделать так `START_BOARD[0][0] = 1;`

Впрочем мы можем закрыть на это глаза раз вы его не пытаетесь изменять. Но по нормальному тут должны быть списки. Например `List.of(-2, -3, -4, -5, -6, -4, -3, -2)`

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/states/UserState.java:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * Состояние пользователя
+     */
+    public enum USER_STATE {
```
Это не константа. Нужно называть так же, как классы

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/states/UserState.java:
```diff
@@ -0,0 +1,99 @@
+        /**
+         * Главное меню
+         */
+        MAINMENU,
```
Нужно использовать SNAKE_CASE

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/states/UserState.java:
```diff
@@ -0,0 +1,99 @@
+     */
+    private final static byte[][] START_BOARD =
+            {
+                    {-2, -3, -4, -5, -6, -4, -3, -2},
```
я уже вам писал замечания про магические числа понятные только вам. Почему вы продолжаете их использовать? В этом массиве не должно быть ни одной цифры, тут должны быть константы с понятными именами

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/chess/Bishop.java:
```diff
@@ -0,0 +1,191 @@
+                    board[startRow - DiagonalShift][startColumn + DiagonalShift] == 0) {
+                possibleMoves.add(new PositionOnBoard(startRow - DiagonalShift,
+                        startColumn + DiagonalShift));
+            } else if (isDownAndLeftFree &&
```
Ребята, учитесь создавать методы. От этого кода начинает кружиться голова, разобраться в нём - невозможно. Зато есть создать методы с понятными именами, то и весь этот метод заиграет новыми красками. Сейчас тут куча внешне одинаковых проверок и действий.

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/chess/Bishop.java:
```diff
@@ -0,0 +1,191 @@
+        List<PositionOnBoard> possibleMoves = new ArrayList<PositionOnBoard>();
+        int startRow = start.row();
+        int startColumn = start.column();
+        int DiagonalShift = 1;
```
переменные нужно называть с маленькой буквы, чтобы никто не путался

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -34,22 +51,164 @@ public TelegramBot(String botToken) {
+     */
+    private void processCallbackQuery(String callbackData, int messageId, long chatId) {
+        List<String> responseMessageText = mainLogic.processInput(callbackData, chatId);
+        Iterator<String> responseMessagesIterator = responseMessageText.iterator();
```
убираем отсюда итератор, вы его используете не по назначению. Вы его используете, чтобы максимально запутать код и сделать так, чтобы никто не понял что вы тут делаете. Комментариев с объяснением тут, разумеется, нет! В общем убираем итератор и начинаем использовать списки, а то сейчас у вас `List<String> responseMessageText` тут просто для красоты, а не для дела.

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -34,22 +51,164 @@ public TelegramBot(String botToken) {
+        sendMessages(chatId, responseMessagesIterator);
+    }
+    
+    private void sendMessages(long chatId, Iterator<String> messagesTextsIterator) {
```
аналогично. Убираем итератор, вы его используете не по назначению. В современных реалиях он нужен крайне редко и то в очень нестандартных случаях, у вас же всё стандартно.

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/TelegramBot.java:
```diff
@@ -34,22 +51,164 @@ public TelegramBot(String botToken) {
+        sendMessages(chatId, responseMessagesIterator);
+    }
+    
+    private void sendMessages(long chatId, Iterator<String> messagesTextsIterator) {
```
нет JavaDoc

vpyzhyanov commented on src/main/java/org/example/chess/GameHandler.java:
```diff
@@ -0,0 +1,143 @@
+    /**
+     * Статус хода
+     */
+    public enum moveProperty {
```
аналогично про название с большой буквы

vpyzhyanov commented on src/main/java/org/example/chess/GameHandler.java:
```diff
@@ -0,0 +1,143 @@
+    /**
+     * Минимальная размерность игрового поля
+     */
+    private int minSideValue = 0;
```
она разве меняется? Тогда она должна быть константой

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/chess/GameHandler.java:
```diff
@@ -0,0 +1,143 @@
+    /**
+     * Максимальная размерность игрового поля
+     */
+    private int maxSideValue = 7;
```
И эта

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/chess/GameHandler.java:
```diff
@@ -0,0 +1,143 @@
+    /**
+     * Получение всех доступных позиций для фигуры в конкретной клетке
+     */
+    public List<PositionOnBoard> allPossiblePositionsForFigure(int figureCode, PositionOnBoard start,
```
неиспользуемый метод в мусорку

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/chess/GameHandler.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    /**
+     * проверка что ход на короля
```
пунктуацию надо поправить

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/chess/Pawn.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Конструктор класса
+     */
+    public Pawn(int minSideValue, int maxSideValue) {
```
То есть вы хотите сказать, что размерность поля постоянно меняется? Зачем вам вообще эти аргументы? И зачем вам изменяемые поля?

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/test/java/org/example/MainLogicTest.java:
```diff
@@ -3,17 +3,135 @@
+	/**
+	 * Обработчик комманд
+	 */
+	private final CommandHandler commandHandler = new CommandHandler();
```
что тут делает этот класс? Вы же собирались тестировать MainLogic

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/test/java/org/example/MainLogicTest.java:
```diff
@@ -3,17 +3,135 @@
+		userStateExpected.setUserState(UserState.USER_STATE.MAINMENU);
+		List<String> responseReal = mainLogic.processInput("/newsinglegame", 0);
+		List<String> responseExpected = 
+				commandHandler.processCommand("newsinglegame", "", userStateExpected);
```
Опять одно и то же замечание. Дублирование логики основного приложения. Я же рассказывал как правильно делать ожидаемые значения. Убираем commandHandler из этого теста.

Раз уж вы пишете интеграционный тест, то не надо делать вид, что он модульный. Пишите его как нормальный интеграционный

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/test/java/org/example/MainLogicTest.java:
```diff
@@ -3,17 +3,135 @@
+	@Test
+	public void userInputCommandTest() {
+		UserState userStateExpected = new UserState();
+		userStateExpected.setUserState(UserState.USER_STATE.MAINMENU);
```
Супер критичная ошибка: в самом начале методы вы вмешиваетесь (буквально хачите) внутренную логику (состояние класса), что приводит к тому, что дальнейшие действия и проверки становятся бессмысленными, ведь вы повлияли на нормальных ход работы класса MainLogic. И так случилось потому, что вы думаете, что тест модульный, но он интеграционный (ну вы его написали именно таким)

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/inputHandlers/MainMenuInputHandler.java:
```diff
@@ -0,0 +1,33 @@
+	
+	@Override
+	public List<String> processInput(String userInput, UserState currentUserState) {
+		if (userInput.equals(NEW_SINGLE_GAME)) {
```
нарушение ООП. За эти проверки отвечает не этот класс, а тот, кто его вызывает. Ну то есть логика должна быть такой: класс занимается своей конкректной работой, а то в какой момент его вызовут - отвечает нет он, а тот, кто его вызвал.

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/test/java/org/example/inputHandlers/MainMenuInputHandlerTest.java:
```diff
@@ -0,0 +1,46 @@
+	@Test
+	public void startGameTest() {
+		UserState userState = new UserState();
+		userState.setUserState(UserState.USER_STATE.MAINMENU);
```
Тоже жёский хак

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/test/java/org/example/inputHandlers/CommandHandlerTest.java:
```diff
@@ -0,0 +1,93 @@
+	 */
+	@Test
+	public void startCommandTest() {
+		userState.setUserState(UserState.USER_STATE.INGAME);
```
во всех тестах одинаковые ошибки. Поэтому тесты ничего не тестируют

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/chess/Queen.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * слон для проверки диагональных ходов
+     */
+    private final Chessmen BISHOP;
```
Внутри королевы слон? О_о Никакая королева вам не простит того, что вы назвали её слоном)))))

Вроде цель благая, вы хотите определить поведение фигуры как комбо из двух фигур, но вот выглядит это очень дико и вы нигде не объяснили суть такого решения, нет ни одного комментария об этом.

KashKetchup replied:
Исправлено

vpyzhyanov commented on src/main/java/org/example/chess/Queen.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * ладья для проверки вертикальных и горизонтальных ходов
+     */
+    private final Chessmen ROOK;
```
это не константа и не надо называть это поле как константу

KashKetchup replied:
Исправлено

vpyzhyanov reviewed (changes requested):
Опять не выполнены базовые требования к проекту

Commit:
Исправлены ошибки

enum теперь называются, как классы.
Убраны лишние static.
Методы Chessmen'ов разбиты на различные методы, убрано дублирование
логики.

Commit:
Убран лишний комментарий

AlexV0rob committed:
Merge pull request #58 from AlexV0rob/task_2_new_fixes

Task 2 new fixes

KashKetchup committed:
Merge remote-tracking branch 'origin/task_2' into task_2

# Conflicts:
#	src/main/java/org/example/GameState.java

KashKetchup committed:
Добавление класса ChessmenMovement

KashKetchup committed:
Измение логики Bishop,Rook и Queen. Queen больше не зависит от Bishop и rook

Commit:
Исправление тестов, избавление от MainMenuInputHandler

Тесты MainLogic, CommandHandler и InGameInputHandler исправлены, тепрь
используют только проверяемые классы и являются полностью модульными.
Убраны изменения состояний напрямую внутри логики класса, теперь
состояние изменяется снаружи.
Работа класса MainMenuInputHandler перенесена в MainLogic, сам класс
удалён за ненадобностью.
Исправлены последние названия enum, где возможно static массивы заменены
на static списки.

KashKetchup committed:
Исправления ошибок в task_2, добавление тестов для

AlexV0rob committed:
Merge pull request #60 from AlexV0rob/task_2_new_fixes

Исправление тестов, избавление от MainMenuInputHandler

KashKetchup committed:
Merge branch 'task_2' into task_2_movement

KashKetchup committed:
Исправление замечаний

KashKetchup committed:
Исправление замечаний

KashKetchup committed:
Merge remote-tracking branch 'origin/task_2_movement' into task_2_movement

KashKetchup committed:
Исправление замечаний, часть 2:
Исправление некорректного сравнения в ChessmenMovement, добавление JavaDoc'ов и тд

KashKetchup committed:
Исправление замечаний, часть 3:
Удаление ненужного метода IsInsideBorders в Rook Bishop и Queen

KashKetchup committed:
Merge pull request #61 from AlexV0rob/task_2_movement

Исправление моей части task_2

AlexV0rob commented:
Список изменений:
1) Поправлена пунктуация, JavaDoc, названия enum и прочая мелочь
2) Итератор заменён на список
3) Поправлены тесты так, чтобы они тестировали только один класс без использования сторонних
4) Исправлена Королева, Ладья и Слон
5) Во всех классов фигур добавлены методы для лучшей читаемости кода

KashKetchup committed:
Удаление ненужного Static в ChessmenMovement

KashKetchup committed:
Merge pull request #65 from AlexV0rob/task_2_ChessmenMovement_fix

Удаление ненужного Static в ChessmenMovement

vpyzhyanov commented on src/main/java/org/example/chess/Queen.java:
```diff
@@ -0,0 +1,63 @@
+    /**
+     * Минимальная размерность игрового поля
+     */
+    private final static int MIN_SIDE_VALUE = 0;
```
не используется почему-то

vpyzhyanov commented on src/main/java/org/example/chess/Queen.java:
```diff
@@ -0,0 +1,63 @@
+    /**
+     * Максимальная размерность игрового поля
+     */
+    private final static int MAX_SIDE_VALUE = 7;
```
не используется почему-то

vpyzhyanov commented on src/main/java/org/example/movement/ChessmenMovement.java:
```diff
@@ -0,0 +1,202 @@
+    private boolean isShiftAvailable(int verticalShift, int horizontalShift,
+                                     int row, int column, byte[][] board) {
+        if (isInsideBorders(row + verticalShift) && isInsideBorders(column + horizontalShift)) {
+            if (isPositionEnemy(row, column, row + verticalShift,
```
2 условия друг в друге легко заменяются на условие `&&`

vpyzhyanov commented on src/test/java/org/example/MainLogicTest.java:
```diff
@@ -2,18 +2,255 @@
+        		List.of(
+        				"Игра началась",
+        				"""
+Ход белых
```
нужно сделать корректный отступ

vpyzhyanov commented on src/test/java/org/example/inputHandlers/CommandHandlerTest.java:
```diff
@@ -0,0 +1,88 @@
+		UserState userState = new UserState();
+		commandHandler.processCommand("newsinglegame", "", userState);
+		Assertions.assertEquals(
+				userState.getUserState(), UserState.UserStatus.IN_GAME);
```
везде перепутаны местами ожадаемое значени и фактическое

vpyzhyanov commented on src/test/java/org/example/inputHandlers/CommandHandlerTest.java:
```diff
@@ -0,0 +1,88 @@
+		commandHandler.processCommand("newsinglegame", "", userState);
+		Assertions.assertEquals(
+				userState.getUserState(), UserState.UserStatus.IN_GAME);
+		commandHandler.processCommand("newsinglegame", "", userState);
```
не проверяется возвращаемый командой ответ

vpyzhyanov commented on src/test/java/org/example/inputHandlers/CommandHandlerTest.java:
```diff
@@ -0,0 +1,88 @@
+		commandHandler.processCommand("newsinglegame", "", userState);
+		commandHandler.processCommand("unknown", "", userState);
+		Assertions.assertEquals(
+				userState.getUserState(), UserState.UserStatus.IN_GAME);
```
вообще-то класс commandHandler не только сменой состояния занимается

vpyzhyanov commented on src/main/java/org/example/movement/ChessmenMovement.java:
```diff
@@ -0,0 +1,202 @@
+/**
+ * Класс, отвественный за перемещение шахматных фигур
+ */
+public class ChessmenMovement {
```
зря вы всё в один класс поместили. Нужно было создать отдельные классы типа "Движение по диагонали", "движение по вертикали и горизонтали" и т.п. Все бы они имени общий интерфейс проверки хода. А уже в фигурах реализовывать поведение фигуры делегируя проверку хода этим классам. А для королевы надо было определить коллекцию этих классов и проверять их с условием ИЛИ.

vpyzhyanov reviewed (approved):
Все критичные замечания исправлены. Остались небольшие замечания. Как исправите, можете мёрджить. Задача принята.

KashKetchup committed:
Разделение ChessmenMovement.java на два класса: VerticalAndHorizontalMovement и DiagonalMovement

KashKetchup committed:
Merge pull request #69 from AlexV0rob/task_2_Movement_reworking

Rework ChessmenMovement
